File: src/main/java/io/xol/chunkstories/api/gui/elements/InputText.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.api.gui.elements;

import io.xol.chunkstories.api.gui.FocusableGuiElement;
import io.xol.chunkstories.api.gui.Font;
import io.xol.chunkstories.api.gui.GuiDrawer;
import io.xol.chunkstories.api.gui.Layer;
import io.xol.chunkstories.api.input.Input;
import io.xol.chunkstories.api.input.Mouse;
import org.joml.Vector4f;

/**
 * Provides a text field to type text into
 */
public class InputText extends FocusableGuiElement {
    private String text = "";

    private Font font;

    private boolean isTransparent = false;
    private boolean password = false;

    public InputText(Layer layer, int x, int y, int width) {
        this(layer, x, y, width, layer.getGui().getFonts().defaultFont());
    }

    public InputText(Layer layer, int x, int y, int width, Font font) {
        super(layer, width, 22);
        setPositionX(x);
        setPositionY(y);
        this.setWidth(width);
        this.setHeight(22);

        this.font = font;
    }

    public boolean handleInput(Input input) {
        if (input.equals("backspace")) {
            if (text.length() > 0)
                text = text.substring(0, text.length() - 1);
            return true;
        }
        return false;
    }

    public boolean handleTextInput(char c) {
        if (c != 0)
            text += c;

        return true;
    }

    public int getWidth() {
        int len = getWidth();
        int txtlen = font.getWidth(text);
        if (txtlen > len)
            len = txtlen;

        return len;
    }

    public boolean isMouseOver(Mouse mouse) {
        return (mouse.getCursorX() >= getPositionX() && mouse.getCursorX() < getPositionX() + getWidth() && mouse.getCursorY() >= getPositionY()
                && mouse.getCursorY() <= getPositionY() + getHeight());
    }

    @Override
    public void render(GuiDrawer renderer) {

        String text = this.text;
        if (password) {
            String passworded = "";
            for (@SuppressWarnings("unused")
                    char c : text.toCharArray())
                passworded += "*";
            text = passworded;
        }

        String backgroundTexture = (isFocused() ? "./textures/gui/textbox.png" : "./textures/gui/textboxnofocus.png");
        if (this.isTransparent)
            backgroundTexture = (isFocused() ? "./textures/gui/textboxnofocustransp.png" : "./textures/gui/textboxtransp.png");

        renderer.drawCorneredBoxTiled(getPositionX(), getPositionY(), getWidth(), getHeight(), 4,
                backgroundTexture, 32);
        renderer.drawStringWithShadow(font, getPositionX() + 4, getPositionY() + 1,
                text + ((isFocused() && System.currentTimeMillis() % 1000 > 500) ? "|" : ""), -1, new Vector4f(1.0f));

    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public boolean isTransparent() {
        return isTransparent;
    }

    public void setTransparent(boolean isTransparent) {
        this.isTransparent = isTransparent;
    }

    public boolean isPassword() {
        return password;
    }

    public void setPassword(boolean password) {
        this.password = password;
    }
}

File: src/main/java/xyz/chunkstories/api/gui/elements/InputText.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.gui.elements

import xyz.chunkstories.api.gui.FocusableGuiElement
import xyz.chunkstories.api.gui.Font
import xyz.chunkstories.api.gui.GuiDrawer
import xyz.chunkstories.api.gui.Layer
import xyz.chunkstories.api.input.Input
import xyz.chunkstories.api.input.Mouse
import org.joml.Vector4f

/**
 * Provides a text field to type text into
 */
class InputText @JvmOverloads constructor(layer: Layer, x: Int, y: Int, width: Int, private val font: Font = layer.gui.fonts.defaultFont()) : FocusableGuiElement(layer, width, 22) {
    var text = ""

    var isTransparent = false
    var isPassword = false

    override var width : Int
        set(value) {super.width = value}
        get() {
            var len = super.width
            val txtlen = font.getWidth(text)
            if (txtlen > len)
                len = txtlen

            return len
        }

    init {
        positionX = x
        positionY = y
        this.width = width
        this.height = 22
    }

    override fun handleInput(input: Input): Boolean {
        if (input.name == "backspace") {
            if (text.length > 0)
                text = text.substring(0, text.length - 1)
            return true
        }
        return false
    }

    fun handleTextInput(c: Char): Boolean {
        if (c.toInt() != 0)
            text += c

        return true
    }

    override fun isMouseOver(mouse: Mouse): Boolean {
        return (mouse.cursorX >= positionX && mouse.cursorX < positionX + width && mouse.cursorY >= positionY
                && mouse.cursorY <= positionY + height)
    }

    override fun render(drawer: GuiDrawer) {

        var text = this.text
        if (isPassword) {
            var passworded = ""
            for (c in text.toCharArray())
                passworded += "*"
            text = passworded
        }

        var backgroundTexture = if (isFocused) "textures/gui/textbox.png" else "textures/gui/textboxnofocus.png"
        if (this.isTransparent)
            backgroundTexture = if (isFocused) "textures/gui/textboxnofocustransp.png" else "textures/gui/textboxtransp.png"

        drawer.drawBoxWithCorners(positionX, positionY, width, height, 8, backgroundTexture)
        drawer.drawStringWithShadow(font, positionX + 4, positionY + 1,
                text + if (isFocused && System.currentTimeMillis() % 1000 > 500) "|" else "", -1, Vector4f(1.0f))

    }
}
